src/exx_evalpao.f90

```diff
@@ -155,19 +155,21 @@ subroutine exx_phi_on_grid(inode,atom,spec,extent,xyz,nsuppfuncs,phi_on_grid,r_i
           pz = pz -ijk(3)+1
        end if overlap_box
        !print*,
+       xyz_offset = xyz + rst
+       !$omp parallel do collapse(3) schedule(runtime) default(none) & 
+       !$omp    shared(mx,my,mz,px,py,pz,grid_spacing,xyz_offset,pao,spec,phi_on_grid,i_dummy,exx_cartesian,extent) &
+       !$omp    private(nx,ny,nz,x,y,z,count1,l1,acz,m1,pao_val)
        grid_x_loop: do nx = mx, px
-          x = xyz(1) + real(nx,double)*grid_spacing + rst(1)
-
           grid_y_loop: do ny = my, py
-             y = xyz(2) + real(ny,double)*grid_spacing + rst(2)
-
              grid_z_loop: do nz = mz, pz
-                z = xyz(3) + real(nz,double)*grid_spacing + rst(3)
+                x = nx*grid_spacing + xyz_offset(1)
+                y = ny*grid_spacing + xyz_offset(2)
+                z = nz*grid_spacing + xyz_offset(3)
```

Comment:
This is fine. If there's spare time, I'd like to try the following:

precompute `x`, `y`, and `z` into arrays outside of the loop. These are just 1d arrays so the memory footprint should be small, and we would avoid redundant recomputations of y and z. Could use a Array of Structures data format so that structures of x,y,z are aligned in memory (ie, `real, dimension(3,N) :: xyz`, where N = max(px-mx, py-my, pz-mz), or something like that. Then inside this loop you could just use `xyz(1,nx)`, `xyz(2,ny)`, `xyz(3,nz)`. I'm not sure if this makes much performance difference. As we learned this week, "memory is expensive, flops are free".

Reply:
I'm not sure this would be worth it. I've just tested the concept with a short program and there seems to be no difference. When accessing xyz in the nested loop, the different values of nx, ny and nz make the memory accessed non-contiguous so we'll be getting lots of cache misses.

Reply:
It should be possible to arrange the data such that the nx,ny,nz accesses are contiguous. Maybe I got it wrong in my comment. In principle I agree, if it seems like this isn't worth it, let's not spend much time on it.